README.md
# FusionAuth Dart Client ![semver 2.0.0 compliant](http://img.shields.io/badge/semver-2.0.0-brightgreen.svg?style=flat-square)
If you're integrating FusionAuth with a Dart or Flutter application, this library will speed up your development time.

For additional information and documentation on FusionAuth refer to [https://fusionauth.io](https://fusionauth.io).

A library for FusionAuth Dart developers.

---

### Known limitations

* The use of some objects are limited due to missing support for dynamic fields (example: custom claims on a jwt). Currently, any data placed in JSON will not render in dart unless the class is explicitly subclassed to contain those fields.
---

## Usage

A simple usage example:

```dart
import 'dart:convert';

import 'package:fusionauth_dart_client/fusionauth_dart_client.dart';

void main() async {
  var client = FusionAuthClient('bf69486b-4733-4470-a592-f1bfce7af580',
      'https://local.fusionauth.io', null);

  var response = await client.searchUsersByQuery(SearchRequest(
      search: UserSearchCriteria(queryString: "email:test@example.com")));

  if (response.successResponse != null) {
    print(json.encode(response.successResponse.users));
  }
}
```

pubspec.yaml:
```yaml
dependencies:
  fusionauth_dart_client: ^1.20.0
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/FusionAuth/fusionauth-dart-client/issues

## Contributors

* [Dmitriy Scherbakov/sherbacov](https://github.com/sherbacov) did auth example. 


CHANGELOG.md
## 1.20.1

- Updated for FusionAuth version 1.20.1

## 1.20.0

- Updated for FusionAuth version 1.20.0

## 1.15.8-beta

- First available version

This version contains the full domain but is limited in requests and responses due to missing support for wildcard getters and settings in the json serialization. Some classes may have issues. This client will remain marked as beta until this is remedied.


test/fusionauth_dart_client_test.dart
import 'package:fusionauth_dart_client/fusionauth_dart_client.dart';
import 'package:test/test.dart';
import 'dart:convert';

import 'dart:io' show Platform;

void main() {
  group('A group of tests', () {
    FusionAuthClient client;

    setUp(() async {
      client = FusionAuthClient(Platform.environment['FUSIONAUTH_API_KEY'] ?? 'bf69486b-4733-4470-a592-f1bfce7af580',
          Platform.environment['FUSIONAUTH_URL'] ?? 'https://local.fusionauth.io', null);

      print('Testing against [${client.host}] with key [${client.apiKey}]');

      var response = await client.searchUsersByQuery(SearchRequest(search: UserSearchCriteria(queryString: "email:test@example.com")));

      if (response.statusCode == 200 && response.successResponse.users != null && response.successResponse.users.isNotEmpty) {
        await client.deleteUser(response.successResponse.users[0].id);
      }
    });

    test('GET', () async {
      var result = await client.retrieveApplications();
      expect(result.statusCode, equals(200));
      expect(result.successResponse, isNotNull);
      expect(result.successResponse.applications.isNotEmpty, isTrue);
    });

    test('POST', () async {
      var request = UserRequest(
          sendSetPasswordEmail: false,
          skipVerification: true,
          user: User(
              email: 'test@example.com'
          )
      );

      // This is a bit uggs. we could probably fix this somehow
      request.user.password = 'password';

      var result = await client.createUser(null, request);

      expect(result.statusCode, equals(200), reason: result.errorResponse != null ? 'Failed because of ${json.encode(result.errorResponse.toJson())}' : '');
      expect(result.successResponse, isNotNull);
      expect(result.successResponse.user.email, equals('test@example.com'));
    });

    // TODO POST, PUT, GET, DELETE, PATCH with json
    // TODO test the toJson of a subclass
    // TODO test an any setter return type
    // TODO test an any setter request type
  });
}


lib/src/Converters.dart
/*
 * Copyright (c) 2020, FusionAuth, All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific
 * language governing permissions and limitations under the License.
 */

import 'package:json_annotation/json_annotation.dart';

import 'FusionAuthDomain.dart';

class IdentityProviderConverter<T> implements JsonConverter<T, dynamic> {
  const IdentityProviderConverter();

  @override
  T fromJson(dynamic json) {
    switch (json['type']) {
      case 'Facebook':
        return FacebookIdentityProvider.fromJson(json) as T;
      case 'Google':
        return GoogleIdentityProvider.fromJson(json) as T;
      case 'Twitter':
        return TwitterIdentityProvider.fromJson(json) as T;
      case 'ExternalJWT':
        return ExternalJWTIdentityProvider.fromJson(json) as T;
      case 'OpenIDConnect':
        return OpenIdConnectIdentityProvider.fromJson(json) as T;
      case 'SAML2':
        return SAMLv2IdentityProvider.fromJson(json) as T;
      default:
        throw UnimplementedError('Missing converter for $T');
    }
  }

  @override
  dynamic toJson(T object) {
    return object;
  }
}

BaseIdentityProvider BaseIdentityProviderFromJson(Map<String, dynamic> json) {
  switch (json['type']) {
    case 'Facebook':
      return FacebookIdentityProvider.fromJson(json);
    case 'Google':
      return GoogleIdentityProvider.fromJson(json);
    case 'Twitter':
      return TwitterIdentityProvider.fromJson(json);
    case 'ExternalJWT':
      return ExternalJWTIdentityProvider.fromJson(json);
    case 'OpenIDConnect':
      return OpenIdConnectIdentityProvider.fromJson(json);
    case 'SAML2':
      return SAMLv2IdentityProvider.fromJson(json);
    default:
      throw UnimplementedError('Missing converter for ${json['type']}');
  }
}

class IdentityProviderApplicationConfigurationConverter<T>
    implements JsonConverter<T, dynamic> {
  const IdentityProviderApplicationConfigurationConverter();

  @override
  T fromJson(dynamic json) {
    if (T is FacebookApplicationConfiguration) {
      return FacebookApplicationConfiguration.fromJson(json) as T;
    } else if (T is GoogleApplicationConfiguration) {
      return GoogleApplicationConfiguration.fromJson(json) as T;
    } else if (T is TwitterApplicationConfiguration) {
      return TwitterApplicationConfiguration.fromJson(json) as T;
    } else if (T is ExternalJWTApplicationConfiguration) {
      return ExternalJWTApplicationConfiguration.fromJson(json) as T;
    } else if (T is OpenIdConnectApplicationConfiguration) {
      return OpenIdConnectApplicationConfiguration.fromJson(json) as T;
    } else if (T is SAMLv2ApplicationConfiguration) {
      return SAMLv2ApplicationConfiguration.fromJson(json) as T;
    } else {
      throw UnimplementedError('Missing converter for $T');
    }
  }

  @override
  dynamic toJson(T object) {
    return object;
  }
}


example/fusionauth_dart_client_auth_example.dart
import 'package:fusionauth_dart_client/fusionauth_dart_client.dart'
    as fusionauth;

void main() async {
  var loginRequest = fusionauth.LoginRequest(
      loginId: 'user@domain.com', password: 'Password!');

  loginRequest.applicationId = '9e0f97c2-5563-4b81-83a8-1db6e6457906';
  loginRequest.noJWT = false;
  loginRequest.metaData = fusionauth.MetaData(
      device: fusionauth.DeviceInfo(
          name: 'IPhone XR', //DeviceName
          type: fusionauth.DeviceType.MOBILE));

  var client = fusionauth.FusionAuthClient(
      'bf69486b-4733-4470-a592-f1bfce7af580', //API Key
      'https://local.fusionauth.io',
      null);

  var result = await client.login(loginRequest);

  if (result.statusCode >= 200 && result.statusCode < 300) {
    print(result.successResponse.toJson());
    if (result.successResponse.user != null) {
      print(result.successResponse.user.toJson());
    }
  } else {
    // error
    print('statusCode: ' + result.statusCode.toString());
    if (result.statusCode == 404) {
      print('Result code 404 can mean the wrong user or password.');
    }

    if (result.errorResponse != null) {
      print(result.errorResponse.toJson());
    }
  }
}


lib/src/FusionAuthRESTClient.dart
/*
 * Copyright (c) 2020, FusionAuth, All Rights Reserved
 */

import 'dart:convert';
import 'dart:io';

import 'FusionAuthDomain.dart';

class ClientResponse<RT, ERT> {
  num statusCode;
  dynamic exception;

  RT successResponse;
  ERT errorResponse;
}

typedef ResponseHandler<RT, ERT> = Future<void> Function(
    HttpClientResponse, ClientResponse<RT, ERT>);

typedef FromJsonMethod<RT> = RT Function(dynamic);

ResponseHandler<RT, ERT> defaultResponseHandlerBuilder<RT, ERT>(
    FromJsonMethod fromJson) {
  return (httpClientResponse, clientResponse) async {
    var body = await httpClientResponse.transform(utf8.decoder).join();
    clientResponse.successResponse = fromJson(json.decode(body));
  };
}

ResponseHandler<RT, ERT> defaultErrorResponseHandlerBuilder<RT, ERT>(
    FromJsonMethod fromJson) {
  return (httpClientResponse, clientResponse) async {
    var body = await httpClientResponse.transform(utf8.decoder).join();
    clientResponse.errorResponse = fromJson(json.decode(body));
  };
}

class FusionAuthRESTClient<ReturnType, ErrorReturnType> {
  dynamic body;
  Map<String, String> headers = Map<String, String>();
  String method;
  Map<String, String> parameters = Map<String, String>();
  String uri;
  HttpClientCredentials credentials;
  String realm;
  String host;
  ResponseHandler<ReturnType, ErrorReturnType> responseHandler;
  ResponseHandler<ReturnType, ErrorReturnType> errorResponseHandler;

  final HttpClient httpClient;

  FusionAuthRESTClient(this.host, this.httpClient) {
    if (ErrorReturnType == Errors) {
      errorResponseHandler = defaultErrorResponseHandlerBuilder((d) => Errors.fromJson(d));
    }
  }

  /// Sets the authorization header using a key
  ///
  /// @param {string} key The value of the authorization header.
  /// @returns {DefaultRESTClient}
  FusionAuthRESTClient withAuthorization(String key) {
    withHeader('Authorization', key);
    return this;
  }

  /// Adds a segment to the request uri
  FusionAuthRESTClient withUriSegment(dynamic segment) {
    if (segment == null) {
      return this;
    }
    uri ??= '';
    if (uri[uri.length - 1] != '/') {
      uri += '/';
    }
    uri = uri + segment;
    return this;
  }

  /// Get the full url + parameter list
  String getFullUrl() {
    return host + uri + _getQueryString();
  }

  /// Sets the body of the client request.
  ///
  /// @param body The object to be written to the request body as form data.
  FusionAuthRESTClient withFormData(Map<String, dynamic> body) {
    this.body = body;
    withHeader('Content-Type', 'application/x-www-form-urlencoded');
    return this;
  }

  /// Adds a header to the request.
  ///
  /// @param key The name of the header.
  /// @param value The value of the header.
  FusionAuthRESTClient withHeader(String key, String value) {
    headers[key] = value;
    return this;
  }

  /// Sets the body of the client request.
  ///
  /// @param body The object to be written to the request body as JSON.
  FusionAuthRESTClient withJSONBody(dynamic body) {
    this.body = body;
    withHeader('Content-Type', 'application/json');
    // Omit the Content-Length, this is set auto-magically by the request library
    return this;
  }

  /// Sets the http method for the request
  FusionAuthRESTClient withMethod(String method) {
    this.method = method;
    return this;
  }

  /// Sets the uri of the request
  FusionAuthRESTClient withUri(String uri) {
    this.uri = uri;
    return this;
  }

  /// Adds parameters to the request.
  ///
  /// @param name The name of the parameter.
  /// @param value The value of the parameter, may be a string, object or number.
  FusionAuthRESTClient withParameter(String name, dynamic value) {
    parameters[name] = value.toString();
    return this;
  }

  /// Sets request's credentials.
  ///
  /// @param value A string indicating whether credentials will be sent with the request always, never, or only when sent to a same-origin URL.
  FusionAuthRESTClient withCredentials(HttpClientCredentials value) {
    credentials = value;
    return this;
  }

  FusionAuthRESTClient withRealm(String realm) {
    this.realm = realm;
    return this;
  }

  FusionAuthRESTClient withResponseHandler(ResponseHandler responseHandler) {
    this.responseHandler = responseHandler;
    return this;
  }

  FusionAuthRESTClient withErrorResponseHandler(
      ResponseHandler errorResponseHandler) {
    this.errorResponseHandler = errorResponseHandler;
    return this;
  }

  /// Run the request and return a promise. This promise will resolve if the request is successful
  /// and reject otherwise.
  Future<ClientResponse<ReturnType, ErrorReturnType>> go() async {
    var clientResponse = ClientResponse<ReturnType, ErrorReturnType>();
    try {
      var request = await _newRequest(Uri.parse(getFullUrl()));

      headers.forEach(request.headers.add);
      if (headers['Content-Type'] == 'application/json') {
        request.write(json.encode(body));
      } else if (headers['Content-Type'] ==
          'application/x-www-form-urlencoded') {
        var sep = '';
        (body as Map).forEach((key, value) {
          request.write(sep +
              Uri.encodeComponent(key) +
              '=' +
              Uri.encodeComponent(value));
          sep = '&';
        });
      } else if (body != null) {
        request.write(body.toString());
      }
      var response = await request.close();

      clientResponse.statusCode = response.statusCode;

      if (response.statusCode == 400 && errorResponseHandler != null) {
        await errorResponseHandler(response, clientResponse);
      }

      if (response.statusCode >= 200 &&
          response.statusCode < 300 &&
          responseHandler != null) {
        await responseHandler(response, clientResponse);
      }
    } catch (e) {
      clientResponse.exception = e;
    }

    return clientResponse;
  }

  Future<HttpClientRequest> _newRequest(Uri uri) {
    if (credentials != null) {
      httpClient.addCredentials(uri, realm, credentials);
    }

    switch (method) {
      case 'GET':
        return httpClient.getUrl(uri);
      case 'DELETE':
        return httpClient.deleteUrl(uri);
      case 'POST':
        return httpClient.postUrl(uri);
      case 'PUT':
        return httpClient.putUrl(uri);
      case 'PATCH':
        return httpClient.patchUrl(uri);
      default:
        throw UnimplementedError('Unimplemented http method');
    }
  }

  String _getQueryString() {
    var queryString = '';
    parameters.forEach((key, value) {
      if (queryString.isEmpty) {
        queryString += '?';
      } else {
        queryString += '&';
      }
      queryString += key + '=' + Uri.encodeComponent(value);
    });
    return queryString;
  }
}
